keps/sig-apps/20200529-delete-priority-annotations.md

```diff
@@ -0,0 +1,291 @@
+## Release Signoff Checklist
+
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This feature allows making a suggestion to the ReplicaSet controller, which pod of a Deployment should be deleted first when a scale-down event happens. This can prevent session disruption in stateful applications in a trivial manner.
+
+## Motivation
+
+For some applications, it is necessary that the application can tell Kubernetes which pod can be deleted and which replica has to be protected. The reason for this is that some applications do have stateful sessions and it is not possible to put such an application into Kubernetes because of session termination resulting from "random" down-scale. If the application is able to tell Kubernetes which of the replicas contains no/few/less important active sessions, this would solve many problems. This feature is non-disruptive to the default behaviour. Only if the annotation is existing, it will make a difference in deletion order.
+
+### Goals
+
+To recommend which pod gets deleted next of a ReplicaSet. This should help to avoid major reworks in existing applications architecture:
+* [45509](https://github.com/kubernetes/kubernetes/issues/45509) - Scale down a deployment by removing specific pods
+
+
+### Non-Goals
+
+Guaranteed (in contrast to the recommendation stated in Goals) deletion of a selected replica.
+
+## Proposal
+
+The application can set the `controller.kubernetes.io/replicaset-downscale-priority` annotation to a pod through the Kubernetes API. When a downscale event happens, the pod with the lower priority value of the previously set annotation will be deleted first. If one pod of the Deployment has no priority annotation set, it will be treated as the lowest priority.
+
+If all pods have the same priority, there is no difference in the normal pod delete decision behaviour. The same applies if the delete priority annotation is not used at all.
```

Comment:
ReplicaSet has its own algorithm to decide which pods to delete first (e.g. delete unhealthy/pending ones before healthy ones). Any pods with the same priority should be deleted using the same algorithm, even if not all of the pods have the same priority.

Reply:
I agree. The delete priority decision has to be checked as a secondary criteria. All important health and assign statuses have a higher importance.

Reply:
>  The delete priority decision has to be checked as a secondary criteria

Would you add this in the KEP as well?

Reply:
Just because saying "secondary" might be confusing, please state the order of all comparisons and how this new feature fits in there.

Reply:
Right, it shouldn't be secondary in general, it actually should be the primary unless there is a strong reason otherwise.

Reply:
looking at the sorting order, it seems there is a concept of rank: https://github.com/kubernetes/kubernetes/blob/4efd56485d10058d69b9f1cd7eb67470c33f12cf/pkg/controller/controller_utils.go#L800

In ReplicaSet the rank is currently used to prioritize deleting pods that share a node (i.e., the rank of a pod equal the number of pods from the same replicaset that exist on the node where the pod was scheduled): https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L814-L825

It seems to me that we could alter the way the rank is calculated in the ReplicaSet to take delete-priority into account (when set, the rank is delete-priority then podsOnNode, when not set it will be just podsOnNode).

Reply:
What if some Pods have delete priority and some don't? Rank and podsOnNode are not necessarily comparable.

Reply:
if some have and some don't, then we need to have a default value for the ones that don't, just like pod priority. 

> Priority and podsOnNode are not necessarily comparable.

They are not, but the rank we pass down should reflect the following order: delete-priority then podsOnNode.

Reply:
But rank is only one number. Are you suggesting it gets replaced by a slice?

Reply:
no, one number is fine, here is what we could do:

Sort the pods based on priority as key1 then on podsOnNode as key2, then iterate over the sorted list to create a rank:

Example: 

- pod1: delete-priority=10, podsOnNode=3
- pod2: delete-priority=1, podsOnNode=1
- pod3: delete-priority=1, podsOnNode=1
- pod4: delete-priority=10, podsOnNode=3
- pod5: delete-priority=5, podsOnNode=2

after sorting it becomes:
- pod2: delete-priority=1, podsOnNode=1
- pod3: delete-priority=1, podsOnNode=1
- pod5: delete-priority=5, podsOnNode=2
- pod1: delete-priority=10, podsOnNode=3
- pod4: delete-priority=10, podsOnNode=3

then assign rank, incrementing when both delete-priority and podsOnNode change while iterating over the sorted list:
- pod2: delete-priority=1, podsOnNode=1      rank=0
- pod3: delete-priority=1, podsOnNode=1      rank=0
- pod5: delete-priority=5, podsOnNode=2     rank=1
- pod1: delete-priority=10, podsOnNode=3    rank=2
- pod4: delete-priority=10, podsOnNode=3    rank=2

Reply:
That's doable, but then we are sorting twice.

Reply:
no, we are sorting once but using two keys

Reply:
just to summarize, we don't need to discuss implementation details here, we only need to say that when delete-priority is set, the rank is delete-priority then podsOnNode, when not set it will be just podsOnNode.